### PR TITLE
Add deterministic state preparation for d<5 CSS codes

### PR DESCRIPTION
**Pull request by lsschmid**
## Description

This PR adds the possibility to synthesize deterministic state preparation protocols for small (d<5) CSS codes.
The resulting protocol can be directly simulated and tested using the Dynamic Subset Sampling from [qsample](https://github.com/dpwinter/qsample).

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: pyproject.toml
Comment (burgholzer) [inline pyproject.toml:60]: Just a quick comment here, because I stumbled over it.
This package hardly seems maintained and is not really well packaged (as is kind of obvious from all the special handling you have to do here).
Do you really need that dependency for the feature you are trying to add here? Or would there be another way to eliminate that dependency and maybe vendor the respective code as part of QECC?

Comment (lsschmid) [reply pyproject.toml:60]: Good point, I already had discussions with Tom on it. Problem is: Without the package most of the code/features will not work (in particular the whole simulation part).
If you prefer one could consider moving the simulation part to an "untested/unimplemented" directory or Zenodo and only keep the construction part which does not rely on qSample.

**Comment by lsschmid** [comment]
Large part of the missing CodeCov (e.g. state_prep_det.py line 547 - 563) are uncovered as they only occur in cases which are feasible to test because of multiple hours of SAT runtime.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Hey :wave:

I just quickly looked through all the changes here and made some very slight modifications.
Notably, I moved the evaluation code outside of the main package directory as it typically does not belong there.

I haven't checked the functionality in detail. I'll leave that up to @pehamTom
All I have left is a hand full of inline comments, which would be nice to address.

File: src/mqt/qecc/ft_stateprep/state_prep_det.py
Comment (pehamTom) [reply src/mqt/qecc/ft_stateprep/state_prep_det.py:15]: In https://github.com/cda-tum/mqt-qecc/pull/275, a lot of stuff has already been refactored since I needed to reuse some of the stateprep methods anyway. It will be a bit before that one is merged, though. I am fine with making `__coset_leader` public for now since it is the easiest solution for @lsschmid right now.

Comment (burgholzer) [reply src/mqt/qecc/ft_stateprep/state_prep_det.py:15]: Note that it is not only this function alone that is being imported, see
https://github.com/cda-tum/mqt-qecc/blob/acd120a3b39d6e24074019ee762a89a024cbd429/src/mqt/qecc/ft_stateprep/state_prep_det.py#L15-L20

Comment (pehamTom) [reply src/mqt/qecc/ft_stateprep/state_prep_det.py:15]: All of these are refactored and made public in the mentioned PR. Making them public in this PR is therefore compatible with my changes.

Comment (burgholzer) [reply src/mqt/qecc/ft_stateprep/state_prep_det.py:15]: Nice. Then let's go with that here.

Comment (lsschmid) [reply src/mqt/qecc/ft_stateprep/state_prep_det.py:15]: I think it makes sense to keep state_prep and state_prep_det in separate files as the code is mostly unrelated to each other. Only point are these helper functions which Tom plans to move to a separate file anyway, so we can do the restructuring there.

Comment (burgholzer) [reply src/mqt/qecc/ft_stateprep/state_prep_det.py:15]: Just to be on the safe side: the functions should still be modified to be public (removing the leading underscores) in this PR here before the whole code can be restructured in a follow-up PR.

File: src/mqt/qecc/ft_stateprep/state_prep.py
Comment (lsschmid) [reply src/mqt/qecc/ft_stateprep/state_prep.py:1101]: This is true and mostly due to the None handling in the existing code. Changing this and making the changes between list and np.array more consistent is part of @pehamTom `s refactoring effort below.

File: src/mqt/qecc/ft_stateprep/state_prep.py
Comment (lsschmid) [reply src/mqt/qecc/ft_stateprep/state_prep.py]: Fixed it, not sure why mypy did not complain about it before but should be fine now.

File: src/mqt/qecc/ft_stateprep/state_prep.py
Comment (burgholzer) [reply src/mqt/qecc/ft_stateprep/state_prep.py:1102]: Alright. In that case, this can be resolved here and need not be part of this PR.

File: src/mqt/qecc/ft_stateprep/state_prep.py
Comment (burgholzer) [reply src/mqt/qecc/ft_stateprep/state_prep.py]: Yeah. I was also kind of surprised by that. It was only PyCharm that was complaining.
Might be worth investigating at some point whether mypy is really set up properly in QECC. For now this can be resolved.

**Review by burgholzer — APPROVED** [review]
Thanks for the additional changes 👍🏼 
I just optimized one of the conversions a little bit and updated the lockfile in the last two commits.

This LGTM for now. Let's get it in 🚀 

@pehamTom when you are refactoring the code, please consider improving on the static typing. I believe it is not fully correct throughout the state prep files, but mypy can't catch it due to Z3 not providing type stubs.
Three further things I noticed:
- as far as I am aware, there is no benefit of using a numpy array over a regular Python list if the list itself stores `bool | z3.BoolRef`. The numpy array simply adds another level of indirection throughout the code. I'd propose to remove that entirely where suitable. This should also simplify quite some cases where one has to check for `np.bool_`
- Most of the functions that are defined throughout the files use very rigid typing for the parameters. Especially when it comes to lists or sets. The Python standard library offers abstract base classes for various concepts (`Iterable`, `Hashable`, etc.) that more generically represent the type of input container that is expected. It is common best practice to keep the typing of parameters as loose as possible, while using a return type that is as narrow as possible.
- Way less important, but it might also be worth to provide (limited) Python stub files for Z3 to get proper typing throughout QECC.

Might be worth to copy this comment over to one of your PRs or Issues. I believe this would improve the code quality quite a bit.

**Pull request merged**
